File: src/net/merayen/elastic/backend/context/MessageHandler.java
package net.merayen.elastic.backend.context;

import java.util.ArrayList;
import java.util.List;

import net.merayen.elastic.backend.context.action.LoadProjectAction;
import net.merayen.elastic.system.intercom.NetListRefreshRequestMessage;
import net.merayen.elastic.system.intercom.ProcessMessage;
import net.merayen.elastic.system.intercom.FinishResetNetListMessage;
import net.merayen.elastic.system.intercom.BeginResetNetListMessage;
import net.merayen.elastic.system.intercom.backend.CreateCheckpointMessage;
import net.merayen.elastic.system.intercom.backend.InitBackendMessage;
import net.merayen.elastic.system.intercom.backend.StartBackendMessage;
import net.merayen.elastic.system.intercom.backend.TidyProjectMessage;
import net.merayen.elastic.util.Postmaster;
import net.merayen.elastic.util.NetListMessages;

/**
 * Helper class for BackendContext.
 * Routes messages between UI, LogicNodes and Processor
 */
public class MessageHandler {
	private final BackendContext backend_context;
	private final Postmaster to_ui = new Postmaster();
	private final Postmaster to_backend = new Postmaster();
	private final Postmaster from_processor = new Postmaster();

	MessageHandler(BackendContext bc) {
		this.backend_context = bc;
	}

	/**
	 * Messages sent from LogicNode further into backend is handled here.
	 */
	void handleFromLogicToProcessor(Postmaster.Message message) {
		backend_context.dispatch.executeMessage(message);
	}

	void handleFromLogicToUI(Postmaster.Message message) {
		to_ui.send(message);
	}

	/**
	 * Handles messages to backend.
	 */
	public void sendToBackend(Postmaster.Message message) {
		to_backend.send(message);
	}

	public void sendToBackend(List<Postmaster.Message> messages) {
		to_backend.send(messages);
	}

	void executeMessagesToBackend() {
		Postmaster.Message message;
		while((message = to_backend.receive()) != null) {
			if(message instanceof NetListRefreshRequestMessage) {
				NetListRefreshRequestMessage m = (NetListRefreshRequestMessage)message;

				List<Postmaster.Message> refresh_messages = new ArrayList<>();
				refresh_messages.add(new BeginResetNetListMessage(m.group_id)); // This will clear the receiver's NetList
				refresh_messages.addAll(NetListMessages.disassemble(backend_context.env.project.getNetList(), m.group_id)); // All these messages will rebuild the receiver's NetList
				refresh_messages.add(new FinishResetNetListMessage());

				to_ui.send(refresh_messages); // Send all messages in a chunk so no other messages can get in-between.

			} else if(message instanceof CreateCheckpointMessage) {
				backend_context.env.project.checkpoint.create();

			} else if(message instanceof TidyProjectMessage) {
				backend_context.env.project.tidy();

			} else if(message instanceof InitBackendMessage) {
				new LoadProjectAction().start(backend_context);

			} else if(message instanceof StartBackendMessage) {
				backend_context.start();

			} else {
				backend_context.logicnode_supervisor.handleMessageFromUI(message);
			}
		}
	}

	public Postmaster.Message[] receiveMessagesFromBackend() {
		return to_ui.receiveAll();
	}

	void queueFromProcessor(Postmaster.Message message) {
		from_processor.send(message);
	}

	void executeMessagesFromProcessor() {
		Postmaster.Message message;
		while((message = from_processor.receive()) != null) {
			if(message instanceof ProcessMessage) {
				backend_context.logicnode_supervisor.handleResponseFromProcessor((ProcessMessage)message);
			} else {
				// XXX handle misc messages from processor? (crash message etc)
			}
		}
	}
}

File: src/main/kotlin/net/merayen/elastic/backend/context/MessageHandler.kt
package net.merayen.elastic.backend.context

import net.merayen.elastic.backend.context.action.ImportFileIntoNodeGroupAction
import net.merayen.elastic.backend.context.action.LoadProjectAction
import net.merayen.elastic.system.intercom.*
import net.merayen.elastic.system.intercom.backend.*
import net.merayen.elastic.util.NetListMessages
import net.merayen.elastic.util.Postmaster
import java.util.*

/**
 * Helper class for BackendContext.
 * Routes messages between UI, LogicNodes and Processor
 */
class MessageHandler internal constructor(private val backendContext: BackendContext) {
    private val to_ui = Postmaster<ElasticMessage>() // FIXME seem to get really big after a good while?
    private val to_backend = Postmaster<ElasticMessage>()
    private val from_processor = Postmaster<ElasticMessage>()

    /**
     * Messages sent from LogicNode further into backend is handled here.
     */
    fun handleFromLogicToProcessor(message: ElasticMessage) {
        backendContext.dispatch.executeMessage(message)
    }

    fun handleFromLogicToUI(message: ElasticMessage) {
        to_ui.send(message)
    }

    fun sendToBackend(messages: Collection<ElasticMessage>) {
        to_backend.send(messages = messages)
    }

    fun executeMessagesToBackend() {
        var message: ElasticMessage?
        while (true) {
            message = to_backend.receive() ?: return

            when (message) {
                is NetListRefreshRequestMessage -> {
                    val m = message as NetListRefreshRequestMessage?

                    val refresh_messages = ArrayList<ElasticMessage>()
                    refresh_messages.add(BeginResetNetListMessage(m!!.group_id)) // This will clear the receiver's NetList
                    refresh_messages.addAll(NetListMessages.disassemble(backendContext.env.project.netList, m.group_id!!)) // All these messages will rebuild the receiver's NetList
                    refresh_messages.add(FinishResetNetListMessage())

                    to_ui.send(refresh_messages) // Send all messages in a chunk so no other messages can get in-between.

                }
                is CreateCheckpointMessage -> backendContext.env.project.checkpoint.create()
                is TidyProjectMessage -> backendContext.env.project.tidy()
                is InitBackendMessage -> LoadProjectAction().start(backendContext)
                is StartBackendMessage -> backendContext.start()
                is ImportFileIntoNodeGroupMessage -> ImportFileIntoNodeGroupAction(message).start(backendContext)
                else -> backendContext.logicnode_supervisor.handleMessageFromUI(message)
            }
        }
    }

    fun receiveMessagesFromBackend(): Collection<ElasticMessage> {
        return to_ui.receiveAll()
    }

    fun queueFromProcessor(message: ElasticMessage) {
        from_processor.send(message)
    }

    fun executeMessagesFromProcessor() {
        var message: ElasticMessage?
        while (true) {
            message = from_processor.receive() ?: return

            when (message) {
                is ProcessMessage -> backendContext.logicnode_supervisor.handleResponseFromProcessor(message as ProcessMessage?)
                else -> {} // XXX handle misc messages from processor? (crash message etc)
            }
        }
    }
}
